sdk/src/trace/batch_span_processor.cc

```diff
@@ -0,0 +1,285 @@
+
+#include "opentelemetry/sdk/trace/batch_span_processor.h"
+#include "src/common/circular_buffer.h"
+
+#include <vector>
+using opentelemetry::sdk::common::AtomicUniquePtr;
+using opentelemetry::sdk::common::CircularBuffer;
+using opentelemetry::sdk::common::CircularBufferRange;
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace trace
+{
+BatchSpanProcessor::BatchSpanProcessor(std::unique_ptr<SpanExporter> &&exporter,
+                                       const size_t max_queue_size,
+                                       const std::chrono::milliseconds schedule_delay_millis,
+                                       const size_t max_export_batch_size)
+    : exporter_(std::move(exporter)),
+      max_queue_size_(max_queue_size),
+      schedule_delay_millis_(schedule_delay_millis),
+      max_export_batch_size_(max_export_batch_size)
+{
+  buffer_ =
+      std::unique_ptr<CircularBuffer<Recordable>>(new CircularBuffer<Recordable>(max_queue_size_));
+
+  // Start the background worker thread
+  worker_thread_ =
+      std::unique_ptr<std::thread>(new std::thread(&BatchSpanProcessor::DoBackgroundWork, this));
+}
+
+std::unique_ptr<Recordable> BatchSpanProcessor::MakeRecordable() noexcept
+{
+  return exporter_->MakeRecordable();
+}
+
+void BatchSpanProcessor::OnStart(Recordable &span) noexcept
+{
+  // no-op
+  (void)span;
+}
+
+void BatchSpanProcessor::OnEnd(std::unique_ptr<Recordable> &&span) noexcept
+{
+  if (is_shutdown_ == true)
+  {
+    // TODO: log in glog
+    return;
+  }
+
+  std::unique_lock<std::mutex> lk(cv_m_);
+
+  if (buffer_->Add(span) == false)
+  {
+    // TODO: glog that spans will likely be dropped
+  }
+
+  // If the queue gets at least half full a preemptive notification is
+  // sent to the worker thread to start a new export cycle.
+  if (buffer_->size() >= max_queue_size_ / 2)
+  {
+    // signal the worker thread
+    cv_.notify_one();
+  }
+}
+
+void BatchSpanProcessor::ForceFlush(std::chrono::microseconds timeout) noexcept
```

Comment:
I think we could also leave timeout support for `ForceFlush` out for now. It makes the code hard to reason about, furthermore we cannot guarantee the timeout (e. g. for concurrent calls to `ForceFlush`).

It is not required by the spec:
> Language library authors can decide if they want to make the flush timeout configurable.

Reply:
I spoke about this with Ilya at length and we decided to use timeout.
What are your thoughts on this, @IlyaKobelevskiy ?

Reply:
This would require us to add timeout to exporter interface, which has already been discussed in https://github.com/open-telemetry/opentelemetry-cpp/issues/203.

Reply:
If timeout is not required by spec, we can drop it here. I think adding timeout to exporter interface is a separate issue.

Reply:
Should I remove it from the processor.h file too or just not use the parameter? @pyohannes

Reply:
Please leave it in the `processor.h`. But add a note to the doc string for the `ForceFlush` for the batched span processor that the timeout functionality is not supported (yet).